migrator/migrator.go

```diff
@@ -0,0 +1,269 @@
+			"rm -rf /var/lib/etcd/member",   // ensure the data folder is empty
+			sedInitialClusterCmd,            // sed command to properly set initialCluster string
+			"systemctl daemon-reload",       // load new etcd3 service file
+			"systemctl start etcd3.service", // restart etcd3, after this etcd3 will start syncing data from the cluster
+		}
+
+		fmt.Printf("Configuring node %s for etcd cluster.\n", nodeName)
+		// execute commands above on the node via k8s job
+		err := m.runCommandsOnNode(nodeName, commands)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	nodeIndex := m.etcdStartingIndex + nodeCount - 1
+	// add the new node to the etcd cluster via etcd client API
+	{
+		peerUrls := []string{etcdPeerName(nodeIndex, m.baseDomain)}
+		r, err := m.etcdClient.Cluster.MemberAdd(ctx, peerUrls)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		fmt.Printf("Added new member %s to the etcd cluster.\n", r.Member.PeerURLs)
+	}
+
+	// wait until k8s api is available again, as etcd data sync will make API unavailable for short time
+	err := waitForApiAvailable(m.k8sClient)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	fmt.Printf("Etcd cluster synced, node %s succesfully joined etcd cluster.\n", nodeNames[nodeIndex])
+
+	return nil
+}
+
+func getMasterNodes(c kubernetes.Interface, labelSelector string) ([]string, error) {
+	var nodeNames []string
+
+	b := backoff.NewMaxRetries(maxRetriesNodes, masterNodeFetchInterval)
+	o := func() error {
+		nodeList, err := c.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: labelSelector})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		if len(nodeList.Items) == masterNodeCount {
+			nodeNames = getNodeNames(nodeList.Items)
+			fmt.Printf("Found %d masters %s, %s, %s.\n", masterNodeCount, nodeNames[0], nodeNames[1], nodeNames[2])
+			return nil
+		} else {
+			fmt.Printf("Found %d masters but expected %d. Retrying in %.2fs\n", len(nodeList.Items), masterNodeCount, masterNodeFetchInterval.Seconds())
+			return microerror.Mask(executionFailedError)
+		}
+	}
+	err := backoff.Retry(o, b)
+	if err != nil {
+		fmt.Printf("Failed to reach k8s API after %d retries.\n", maxRetriesApi)
+		return nil, microerror.Mask(err)
+	}
+	return nodeNames, nil
+}
+
+// waitForApiAvailable wait until k8s api is available which indicates that etcd cluster is synced with the new member.
+func waitForApiAvailable(c kubernetes.Interface) error {
+	fmt.Printf("Waiting for the etcd data sync.\n")
+	time.Sleep(waitApiStartInterval)
+
+	b := backoff.NewMaxRetries(maxRetriesApi, waitApiRetryInterval)
+	o := func() error {
+		_, err := c.CoreV1().Nodes().List(metav1.ListOptions{})
+		if err != nil {
+			fmt.Printf("API is still down. retrying in %.2f.\n", waitApiRetryInterval.Seconds())
+			return microerror.Mask(err)
+		}
+
+		return nil
+	}
+	err := backoff.Retry(o, b)
+	if err != nil {
+		fmt.Printf("Failed to reach k8s API after %d retries.\n", maxRetriesApi)
+		return microerror.Mask(err)
```

Comment:
What happens in this case? Is the job started again or has the migration failed at this point?

Reply:
the pod will exit with an error which will make the  app start  again and than it depends if the k8s api works or not.

Theoretically it can mean the migration failed completely, but its hard to guess at this point without some global testing or failures